esp/esdllib/esdl_script.cpp

```diff
@@ -226,6 +338,152 @@ class CEsdlTransformOperationSetValue : public CEsdlTransformOperationBase
     }
 };
 
+class CEsdlTransformOperationRenameNode : public CEsdlTransformOperationBase
+{
+protected:
+    StringAttr m_target;
+    StringAttr m_new_name;
+    Owned<ICompiledXpath> m_xpath_target;
+    Owned<ICompiledXpath> m_xpath_new_name;
+
+public:
+    CEsdlTransformOperationRenameNode(IPropertyTree *tree, const StringBuffer &prefix) : CEsdlTransformOperationBase(tree, prefix)
+    {
+        if (isEmptyString(tree->queryProp("@new_name")) && isEmptyString(tree->queryProp("@xpath_new_name")))
+            esdlOperationError(ESDL_SCRIPT_MissingOperationAttr, m_tagname.str(), "without new name", m_traceName.str(), !m_ignoreCodingErrors);
+
+        if (isEmptyString(tree->queryProp("@target")) && isEmptyString(tree->queryProp("@xpath_target")))
+            esdlOperationError(ESDL_SCRIPT_MissingOperationAttr, m_tagname.str(), "without target", m_traceName.str(), !m_ignoreCodingErrors);
+
+        if (tree->hasProp("@xpath_target"))
+            m_xpath_target.setown(compileXpath(tree->queryProp("@xpath_target")));
+        else if (tree->hasProp("@target"))
+            m_target.set(tree->queryProp("@target"));
+
+        if (tree->hasProp("@xpath_new_name"))
+            m_xpath_new_name.setown(compileXpath(tree->queryProp("@xpath_new_name")));
+        else if (tree->hasProp("@new_name"))
+            m_new_name.set(tree->queryProp("@new_name"));
+    }
+
+    virtual void toDBGLog() override
+    {
+#if defined(_DEBUG)
+        const char *target = (m_xpath_target) ? m_xpath_target->getXpath() : m_target.str();
+        const char *new_name = (m_xpath_new_name) ? m_xpath_new_name->getXpath() : m_new_name.str();
+        DBGLOG(">%s> %s(%s, new_name('%s'))", m_traceName.str(), m_tagname.str(), target, new_name);
+#endif
+    }
+
+    virtual ~CEsdlTransformOperationRenameNode(){}
+
+    virtual bool process(IEsdlScriptContext * context, IPropertyTree *content, IXpathContext * xpathContext) override
+    {
+        if ((!m_xpath_target && m_target.isEmpty()) || (!m_xpath_new_name && m_new_name.isEmpty()))
+            return false; //only here if "optional" backward compatible support for now (optional syntax errors aren't actually helpful
+        try
+        {
+            StringBuffer path;
+            if (m_xpath_target)
+                xpathContext->evaluateAsString(m_xpath_target, path);
+            else
+                path.set(m_target);
+
+            StringBuffer name;
+            if (m_xpath_new_name)
+                xpathContext->evaluateAsString(m_xpath_new_name, name);
+            else
+                name.set(m_new_name);
+
+            //make sure we only remove one.  we can add a multi-node support in the future (when we are libxml2 based?)
+            //don't want users to depend on quirky behavior
+            path.append("[1]"); //don't need to check as multiple [] are ok
```

Comment:
With the libxml2 version of this, my intent was to pass an attribute specifying if all occurrences were to be affected. If not applied to all, a multiple node nodeset would be considered a failure. Just picking one of many seems like quirky behavior best to be avoided.

Reply:
I can throw an exception if multiple... and leave it to the script writer to add the [1] if desired.  There are cases where since this state can creep up externally that causing the whole transaction to fail could be undesirable.  This will mean script writers need to be diligent of this at all times.